# 00-02 - Variables and Values

In 00-01 we gave a name to a number and saw that the kernel remembered it. Here we take that seriously. We look at what kinds of thing Python can hold, what the difference between them is, and why the difference matters more than it first appears.

Everything we count in this notebook is real. Princeton has 7893 mapped buildings, 92 places to eat, and an area of 47.66 square kilometres, and by the end of this notebook you will have combined those into something nobody handed you.

## Giving something a name

Python lets you attach a name to a value, which is called **assignment**. The name goes on the left, the value on the right:

In [1]:
buildings = 7893

Nothing printed, which is correct: we stored the value rather than asking for it.

The equals sign here is worth pausing on, because it does not mean what it means in mathematics. In mathematics `x = 5` is a statement of fact that is either true or false. In Python it is an **instruction**, and it points right to left: take the value on the right, and let the name on the left refer to it from now on.

That difference has a consequence that looks absurd on paper and is completely ordinary in code:

In [2]:
buildings = buildings + 1
print(buildings)

7894


7894. As a mathematical statement that line is nonsense, since nothing equals itself plus one. As an instruction it is clear: work out what is on the right, i.e. 7893 plus 1, then let `buildings` refer to that.

Let us put it back:

In [3]:
buildings = 7893
print(buildings)

7893


Note that a name can be reused freely, and the old value is simply forgotten. Nothing warns you about this, which is occasionally a problem: if you use the same name for two different things in one notebook, the second one wins from that point down.

**Choosing names.** They may contain letters, digits and underscores, but may not begin with a digit and may not contain spaces or hyphens. `area_km2` is fine, `2nd_area` and `area km2` are not. Capital letters matter, so `Buildings` and `buildings` are two different names.

Write names for the human who reads them, and that human is usually you in three weeks. `n` is faster to type than `buildings`, and much slower to understand.

## Kinds of value

Python cares what kind of thing a value is, and there are four kinds you will meet constantly.

A **whole number**, called an integer or `int`, like our count of buildings. A **number with a decimal point**, called a float, like an area. A piece of **text**, called a string or `str`, always written inside quotes. And a **true or false** value, called a boolean or `bool`, written `True` and `False` with capital letters.

The function `type` reports which one you have:

In [4]:
buildings = 7893
area_km2 = 47.66
town = "Princeton"
is_in_new_jersey = True

print(type(buildings))
print(type(area_km2))
print(type(town))
print(type(is_in_new_jersey))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


Four lines, reading `int`, `float`, `str` and `bool`.

Note the quotes around `"Princeton"`. They are not decoration and they are not optional. They are how Python knows that what follows is text to be stored rather than a name to be looked up. Without them, Python would go looking for something called `Princeton` and complain that it has never heard of it, which is the `NameError` from 00-01.

## Why the kind matters

Here is where beginners lose an hour, so let us lose thirty seconds instead.

The number `7893` and the text `"7893"` look almost identical on the page. They behave completely differently, because one is a quantity and the other is four characters that happen to be digits.

Watch:

In [5]:
"7893" + 1

TypeError: can only concatenate str (not "int") to str

`TypeError: can only concatenate str (not "int") to str`.

Python is saying that `+` between two pieces of text means gluing them together, and you gave it a number to glue on, and it will not guess whether you meant to count or to glue.

The fix is to say which you meant. `int()` turns text into a whole number:

In [6]:
print(int("7893") + 1)
print("7893" + "1")

7894
78931


7894 and `78931`. Both ran without complaint, and only one of them is arithmetic. Note that the second one produced text, not a number, even though it looks like one.

This matters far beyond arithmetic. Data you read from a file often arrives as text when you expected numbers, particularly if a stray character crept into a column, and then everything downstream behaves strangely without anything failing. Knowing to check `type` is the cheap fix.

## Combining values

Now let us make something nobody handed us. We have a count of buildings and an area, so we can work out how densely built Princeton is:

In [7]:
buildings_per_km2 = buildings / area_km2
print(buildings_per_km2)

165.61057490558122


About 165.6.

Two things to notice. First, dividing two numbers gave us a float even though `buildings` was an integer, because Python does not silently throw away the fractional part.

Second, the answer has far more decimal places than we have any right to. Our building count comes from volunteers who mapped what they felt like mapping, and the boundary is one institution's claim about where the town ends. Reporting 165.6193 buildings per square kilometre would suggest a precision that does not exist.

## Putting values into a sentence

Printing bare numbers is fine for checking. For anything a human reads, you want the number inside a sentence, and the tool for that is an **f-string**: an ordinary string with the letter `f` in front of the opening quote, where anything inside curly braces gets worked out and dropped in.

In [8]:
print(f"{town} has {buildings} mapped buildings in {area_km2} square kilometres.")

Princeton has 7893 mapped buildings in 47.66 square kilometres.


The names inside the braces were replaced by their values, and the rest of the text came through as written.

You can also say how a number should look, by putting a colon and a format after the name. `:.0f` means a float with no decimal places:

In [9]:
print(f"That is about {buildings_per_km2:.0f} buildings per square kilometre.")

That is about 166 buildings per square kilometre.


About 166 buildings per square kilometre, which is a number you could say out loud to somebody.

The formats worth knowing are `:.0f` and `:.2f` for no or two decimal places, and `:,` which puts thousands separators in, so `f"{buildings:,}"` gives `7,893`.

Note that the rounding happens only in the display. The value in `buildings_per_km2` is unchanged, and if you use it again you get the full number. That is usually what you want: round when you show, never while you calculate.

## Comments

A line beginning with `#` is ignored by Python entirely. It is there for people:

In [10]:
# Volunteers map what interests them, so this is a floor rather than a count.
print(f"{buildings:,} buildings")

7,893 buildings


The comment did not appear in the output, and it was not executed.

Note what a good comment is for. It should say **why**, not what. A comment reading `# divide buildings by area` is noise, because the line below it already says that. A comment saying why the number is a floor rather than a count tells the reader something the code cannot.

## Check your understanding

1. What is in `count` after these two lines run: `count = 92` then `count = count + 1`?
2. `type(47.66)` returns something. What?
3. Why does `"92" + "1"` give `921` rather than 93, and what would you write to get 93?
4. What is the difference between `print(f"{area_km2:.1f}")` and changing the value of `area_km2` itself?

## Where we are

You can store a value, you know the four kinds you will meet, you know why the kind matters, and you can put a number into a sentence a human would read.

In 00-03 we look at the other half of every line of Python you will write: calling a function, and understanding what its arguments do.

## Further resources

Nothing in this course requires anything below.

The official Python tutorial covers numbers and strings in its third chapter: https://docs.python.org/3/tutorial/introduction.html. The full grammar of f-string formats, if you ever need an unusual one, is at https://docs.python.org/3/library/string.html#format-specification-mini-language.

The counts used here describe an OpenStreetMap extract of Princeton, snapshot 2026-08-29, licensed ODbL 1.0, copyright OpenStreetMap contributors: https://www.openstreetmap.org/copyright.